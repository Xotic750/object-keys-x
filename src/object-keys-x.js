import toObject from 'to-object-x';
import attempt from 'attempt-x';
import isArray from 'is-array-x';
import isArguments from 'is-arguments';
import arraySlice from 'array-like-slice-x';
import splitIfBoxed from 'split-if-boxed-bug-x';
import isString from 'is-string';
import isRegexp from 'is-regexp-x';
import has from 'has-own-property-x';
import objKeys from 'object-keys';

const ObjectCtr = {}.constructor;
const nativeKeys = typeof ObjectCtr.keys === 'function' && ObjectCtr.keys;

let isWorking;
let throwsWithNull;
let worksWithPrim;
let worksWithRegex;
let worksWithArgs;
let worksWithStr;

if (nativeKeys) {
  const isCorrectRes = function isCorrectRes(r, length) {
    return r.threw === false && isArray(r.value) && r.value.length === length;
  };

  const either = function either(r, a, b) {
    const x = r.value[0];
    const y = r.value[1];

    return (x === a && y === b) || (x === b && y === a);
  };

  let testObj = {a: 1, b: 2};
  let res = attempt(nativeKeys, testObj);
  isWorking = isCorrectRes(res, 2) && either(res, 'a', 'b');

  if (isWorking) {
    testObj = Object('a');
    testObj.y = 1;
    res = attempt(nativeKeys, testObj);
    isWorking = isCorrectRes(res, 2) && either(res, '0', 'y');
  }

  if (isWorking) {
    throwsWithNull = attempt(nativeKeys, null).threw;
    worksWithPrim = isCorrectRes(attempt(nativeKeys, 42), 0);
    worksWithRegex = attempt(nativeKeys, /a/g).threw === false;

    res = attempt(
      nativeKeys,
      (function getArgs() {
        /* eslint-disable-next-line prefer-rest-params */
        return arguments;
      })(1, 2),
    );

    worksWithArgs = isCorrectRes(res, 2) && either(res, '0', '1');

    res = attempt(nativeKeys, Object('ab'));
    worksWithStr = isCorrectRes(res, 2) && either(res, '0', '1');
  }
}

export const patched = function keys(object) {
  let obj = toObject ? toObject(object) : object;

  if (worksWithArgs !== true && isArguments(obj)) {
    obj = arraySlice(obj);
  } else if (worksWithStr !== true && isString(obj)) {
    obj = splitIfBoxed(obj);
  } else if (worksWithRegex !== true && isRegexp(obj)) {
    const regexKeys = [];
    /* eslint-disable-next-line no-restricted-syntax */
    for (const key in obj) {
      // noinspection JSUnfilteredForInLoop
      if (has(obj, key)) {
        regexKeys[regexKeys.length] = key;
      }
    }

    return regexKeys;
  }

  return nativeKeys(obj);
};

export const implementation = function keys(object) {
  return objKeys(toObject(object));
};

let objectKeys;

if (isWorking) {
  if (throwsWithNull && worksWithPrim && worksWithRegex && worksWithArgs && worksWithStr) {
    objectKeys = nativeKeys;
  } else {
    objectKeys = patched;
  }
}

/**
 * This method returns an array of a given object's own enumerable properties,
 * in the same order as that provided by a for...in loop (the difference being
 * that a for-in loop enumerates properties in the prototype chain as well).
 *
 * @param {*} obj - The object of which the enumerable own properties are to be returned.
 * @returns {Array} An array of strings that represent all the enumerable properties of the given object.
 */
const $objectKeys = isWorking ? objectKeys : implementation;

export default $objectKeys;
