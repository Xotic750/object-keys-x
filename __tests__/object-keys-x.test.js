import $A, {implementation as $I, patched as $P} from '../src/object-keys-x';

const has = Object.prototype.hasOwnProperty;
const supportsDescriptors =
  Object.defineProperty &&
  (function() {
    try {
      const obj = {};
      Object.defineProperty(obj, 'x', {enumerable: false, value: obj});
      /* eslint-disable-next-line no-restricted-syntax,guard-for-in */
      for (const _ in obj) {
        return false;
      }

      return obj.x === obj;
    } catch (e) {
      /* this is ES3 */
      return false;
    }
  })();

const ifWindowIt = typeof window === 'undefined' ? xit : it;

[$I, $P, $A].forEach((objectKeys, testNum) => {
  describe(`objectKeys ${testNum}`, function() {
    const obj = {
      arr: [],
      bool: true,
      null: null,
      num: 42,
      obj: {},
      str: 'boz',

      undefined: void 0,
    };

    const loopedValues = [];
    /* eslint-disable-next-line no-restricted-syntax,guard-for-in */
    for (const key in obj) {
      loopedValues.push(key);
    }

    const keys = objectKeys(obj);
    it('should throw for null or undefined', function() {
      expect.assertions(3);
      expect(function() {
        objectKeys();
      }).toThrowErrorMatchingSnapshot();

      expect(function() {
        objectKeys(void 0);
      }).toThrowErrorMatchingSnapshot();

      expect(function() {
        objectKeys(null);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should not throw for non-objects', function() {
      expect.assertions(3);
      expect(objectKeys(1)).toStrictEqual([]);
      expect(objectKeys(true)).toStrictEqual([]);
      expect(objectKeys('')).toStrictEqual([]);
    });

    it('should have correct length', function() {
      expect.assertions(1);
      expect(keys).toHaveLength(7);
    });

    describe('arguments objects', function() {
      it('works with an arguments object', function() {
        expect.assertions(3);
        (function() {
          expect(arguments).toHaveLength(3);

          expect(objectKeys(arguments)).toHaveLength(arguments.length);

          expect(objectKeys(arguments)).toStrictEqual(['0', '1', '2']);
        })(1, 2, 3);
      });

      it('works with a legacy arguments object', function() {
        expect.assertions(1);
        const FakeArguments = function(args) {
          args.forEach(
            function(arg, i) {
              this[i] = arg;
            }.bind(this),
          );
        };

        FakeArguments.prototype.length = 3;

        FakeArguments.prototype.callee = function() {};

        const fakeOldArguments = new FakeArguments(['a', 'b', 'c']);
        expect(objectKeys(fakeOldArguments)).toStrictEqual(['0', '1', '2']);
      });
    });

    it('should return an Array', function() {
      expect.assertions(1);
      expect(Array.isArray(keys)).toBe(true);
    });

    it('should return names which are own properties', function() {
      expect.assertions(7);
      keys.forEach(function(name) {
        expect(has.call(obj, name)).toBe(true);
      });
    });

    it('should return names which are enumerable', function() {
      expect.assertions(7);
      keys.forEach(function(name) {
        expect(loopedValues.indexOf(name)).not.toBe(-1);
      });
    });

    describe('enumerating over non-enumerable properties', function() {
      it('has no enumerable keys on a Function', function() {
        expect.assertions(1);

        const Foo = function() {};

        expect(objectKeys(Foo.prototype)).toStrictEqual([]);
      });

      it('has no enumerable keys on a boolean', function() {
        expect.assertions(1);
        expect(objectKeys(Boolean.prototype)).toStrictEqual([]);
      });

      it('has no enumerable keys on an object', function() {
        expect.assertions(1);
        expect(objectKeys(Object.prototype)).toStrictEqual([]);
      });
    });

    it('works with boxed primitives', function() {
      expect.assertions(1);
      expect(objectKeys(Object('hello'))).toStrictEqual(['0', '1', '2', '3', '4']);
    });

    it('works with boxed primitives with extra properties', function() {
      expect.assertions(1);
      const x = Object('x');
      x.y = 1;
      const actual = objectKeys(x);
      const expected = ['0', 'y'];
      actual.sort();
      expected.sort();
      expect(actual).toStrictEqual(expected);
    });

    it('works with regexs', function() {
      expect.assertions(1);
      const x = /a/g;
      const actual = objectKeys(x).sort();
      expect(actual).toStrictEqual([]);
    });

    ifWindowIt('can serialize all objects on the `window`', function() {
      expect.assertions(62);
      let windowItemKeys;
      let exception;
      const excludedKeys = [
        'window',
        'console',
        'parent',
        'self',
        'frame',
        'frames',
        'frameElement',
        'external',
        'height',
        'width',
      ];

      if (supportsDescriptors) {
        Object.defineProperty(window, 'thrower', {
          configurable: true,
          get() {
            throw new RangeError('thrower!');
          },
        });
      }

      /* eslint-disable-next-line guard-for-in,no-restricted-syntax */
      for (const k in window) {
        exception = void 0;
        windowItemKeys = exception;

        if (excludedKeys.indexOf(k) === -1 && has.call(window, k) && window[k] !== null && typeof window[k] === 'object') {
          try {
            windowItemKeys = objectKeys(window[k]);
          } catch (e) {
            exception = e;
          }

          expect(Array.isArray(windowItemKeys)).toBe(true);
          expect(exception).toBeUndefined();
        }
      }

      if (supportsDescriptors) {
        delete window.thrower;
      }
    });
  });
});
