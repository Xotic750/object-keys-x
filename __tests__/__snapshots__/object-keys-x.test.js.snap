// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`objectKeys 0 should throw for null or undefined 1`] = `"Cannot call method on undefined"`;

exports[`objectKeys 0 should throw for null or undefined 2`] = `"Cannot call method on undefined"`;

exports[`objectKeys 0 should throw for null or undefined 3`] = `"Cannot call method on null"`;

exports[`objectKeys 1 should throw for null or undefined 1`] = `"Cannot call method on undefined"`;

exports[`objectKeys 1 should throw for null or undefined 2`] = `"Cannot call method on undefined"`;

exports[`objectKeys 1 should throw for null or undefined 3`] = `"Cannot call method on null"`;

exports[`objectKeys 2 should throw for null or undefined 1`] = `"Cannot convert undefined or null to object"`;

exports[`objectKeys 2 should throw for null or undefined 2`] = `"Cannot convert undefined or null to object"`;

exports[`objectKeys 2 should throw for null or undefined 3`] = `"Cannot convert undefined or null to object"`;
